File: 1188-brace-expansion-ii/brace-expansion-ii.py
class Solution:
    def braceExpansionII(self, expression):
        def parse(s):
            res = {""}
            i = 0

            while i < len(s):
                if s[i] == '{':
                    j = i
                    count = 0

                    while j < len(s):
                        if s[j] == '{':
                            count += 1
                        elif s[j] == '}':
                            count -= 1

                        if count == 0:
                            break
                        j += 1

                    parts = split(s[i + 1:j])
                    res = {a + b for a in res for b in parts}
                    i = j + 1

                elif s[i].isalpha():
                    res = {x + s[i] for x in res}
                    i += 1

            return res

        def split(s):
            res = []
            start = 0
            count = 0

            for i, c in enumerate(s):
                if c == '{':
                    count += 1
                elif c == '}':
                    count -= 1
                elif c == ',' and count == 0:
                    res.extend(parse(s[start:i]))
                    start = i + 1

            res.extend(parse(s[start:]))
            return res

        return sorted(parse(expression))

File: 1188-brace-expansion-ii/brace-expansion-ii.java
class Solution {
    public List<String> braceExpansionII(String expression) {
        Set<String> set = new TreeSet<>();
        dfs(expression, 0, set);
        return new ArrayList<>(set);
    }

    void dfs(String s, int idx, Set<String> set) {
        if (idx == s.length()) {
            set.add(s);
            return;
        }

        if (s.charAt(idx) == '{') {
            int end = idx, count = 0;
            while (end < s.length()) {
                if (s.charAt(end) == '{') count++;
                if (s.charAt(end) == '}') count--;
                if (count == 0) break;
                end++;
            }

            String inside = s.substring(idx + 1, end);
            List<String> parts = split(inside);

            for (String p : parts) {
                dfs(s.substring(0, idx) + p + s.substring(end + 1), idx, set);
            }
        } else {
            dfs(s, idx + 1, set);
        }
    }

    List<String> split(String s) {
        List<String> res = new ArrayList<>();
        int count = 0, start = 0;

        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '{') count++;
            else if (s.charAt(i) == '}') count--;
            else if (s.charAt(i) == ',' && count == 0) {
                res.add(s.substring(start, i));
                start = i + 1;
            }
        }

        res.add(s.substring(start));
        return res;
    }
}